### PR TITLE
Invalid synced tips

### PR DESCRIPTION
**Pull request by potuz**
This PR handles the situation when a node changes status SYNCING -> INVALID

1- removes the invalid node from fork choice
2- readjusts children/parent relations
3- takes into account weight to readjust best descendant and child
4- updates the synced tips structures to reflect this change.

This last case 4 can only happen in one scenario:

```
0 -- 1 -- 2
        \ 
          -- 3
```
if 3 becomes INVALID, then

a- if 1 was not a synced tip, then there are no changes
b- is 1 was a synced tip only because it was the tip of the branch ending in 3 then it has to be removed. For example, if 2 was a synced tip the 1 needs to be removed, if 2 was not a synced tip, then 1 continues being a synced tip.

### CODE REVIEW

File: beacon-chain/forkchoice/protoarray/optimistic_sync.go [hunk 1]
```diff
@@ -88,3 +88,204 @@ func (f *ForkChoice) Optimistic(ctx context.Context, root [32]byte, slot types.S
...
+	// Stop early if the root is part of validated tips
+	f.syncedTips.Lock()
+	defer f.syncedTips.Unlock()
+	_, ok = f.syncedTips.validatedTips[root]
+	if ok {
+		return nil
+	}
+
+	// Cache root and slot to validated tips
+	f.syncedTips.validatedTips[root] = node.slot
```
Comment (prestonvanloon) [inline beacon-chain/forkchoice/protoarray/optimistic_sync.go:117]: Shouldn't you do this _after_ the validation?

What if the context times out before validation has been completed?

Comment (potuz) [reply beacon-chain/forkchoice/protoarray/optimistic_sync.go:117]: Oh sorry it took me a little to understand this comment is about a different PR. So this method takes a root of a block that we already know from the EL that it's VALID, since it's a tip of the fork choice tree then it has to be a synced tip no matter what

File: beacon-chain/forkchoice/protoarray/optimistic_sync.go [hunk 2]
```diff
@@ -88,3 +88,204 @@ func (f *ForkChoice) Optimistic(ctx context.Context, root [32]byte, slot types.S
...
+	// parent's best child and descendant.
+	needsUpdateParentsChildren := false
+	if parent.bestChild == idx || parent.bestDescendant == idx {
+		needsUpdateParentsChildren = true
+	}
+
+	// delete the invalid node, order is important
+	f.store.nodes = append(f.store.nodes[:idx], f.store.nodes[idx+1:]...)
+	delete(f.store.nodesIndices, root)
+	// Fix best parent and best child for each node
```
Comment (terencechain) [inline beacon-chain/forkchoice/protoarray/optimistic_sync.go:223]:
```suggestion
	// Fix parent and best child for each node
```

there's just one parent per node

File: beacon-chain/forkchoice/protoarray/optimistic_sync.go [hunk 3]
```diff
@@ -88,3 +88,204 @@ func (f *ForkChoice) Optimistic(ctx context.Context, root [32]byte, slot types.S
...
+	if !ok {
+		return errInvalidNodeIndex
+	}
+	node := f.store.nodes[idx]
+	parentIndex := node.parent
+
+	f.syncedTips.Lock()
+	defer f.syncedTips.Unlock()
+
+	// This should not happen
```
[the remark this answers is not in the document]
Comment (potuz) [reply beacon-chain/forkchoice/protoarray/optimistic_sync.go:206]: Indeed, but in that case what should we do? we're synced without forks and the head is invalid.

[the remark this answers is not in the document]
Comment (terencechain) [reply beacon-chain/forkchoice/protoarray/optimistic_sync.go:207]: Just remove the head? We probably should add a test case for it too

File: beacon-chain/forkchoice/protoarray/optimistic_sync.go [hunk 4]
```diff
@@ -89,17 +89,17 @@ func (f *ForkChoice) Optimistic(ctx context.Context, root [32]byte, slot types.S
 	return f.Optimistic(ctx, root, slot)
 }
 
-// UpdateSyncedTips updates the synced_tips map when the block with the given root becomes VALID
-func (f *ForkChoice) UpdateSyncedTips(ctx context.Context, root [32]byte) error {
+// UpdateSyncedTip updates the synced_tips map when the block with the given root becomes VALID
+func (f *ForkChoice) UpdateSyncedTip(ctx context.Context, root [32]byte) error {
```
Comment (potuz) [inline beacon-chain/forkchoice/protoarray/optimistic_sync.go:93]: Why this change, we are updating all the syncedTips not just one

File: beacon-chain/forkchoice/protoarray/optimistic_sync.go [hunk 5]
```diff
@@ -188,17 +187,16 @@ func (f *ForkChoice) UpdateSyncedTips(ctx context.Context, root [32]byte) error
 	return nil
 }
 
-// This function removes a node that has become INVALID according to the
-// EL. It updates the synced_tips to reflect this change.
-func (f *ForkChoice) Invalid(ctx context.Context, root [32]byte) error {
+// RemoveSyncTip removes a node with root from Fork Choice store. It updates the synced tips map.
+func (f *ForkChoice) RemoveSyncTip(ctx context.Context, root [32]byte) error {
```
Comment (potuz) [inline beacon-chain/forkchoice/protoarray/optimistic_sync.go:191]: This is a bit misleading since this function may not remove any synced tip, take for example 

```
0  -- 1 
 \ 
   --  2
```
with 0 being the unique synced tip and 1 and 2 optimistic. If 1 becomes invalid, it is removed, but synced tips do not change

Comment (terencechain) [reply beacon-chain/forkchoice/protoarray/optimistic_sync.go:191]: I'll think of better names for them. I mainly want to express some form of symmetric for the two functions

File: beacon-chain/forkchoice/protoarray/optimistic_sync.go [hunk 6]
```diff
@@ -245,9 +232,8 @@ func (f *ForkChoice) Invalid(ctx context.Context, root [32]byte) error {
 		}
 	}
 
-	// Update the parent's best child and best descendant
-	if needsUpdateParentsChildren {
-		// look for a new child
+	// Update the parent's best child and best descendant if necessary.
+	if parent.bestChild == idx || parent.bestDescendant == idx {
```
Comment (potuz) [inline beacon-chain/forkchoice/protoarray/optimistic_sync.go:236]: Are you sure this is safe? the reason I had the boolean up above is that we change the parent/child indices in the loop above this one. `idx` does not mean the same that it used to mean above.

**Review by terencechain — APPROVED** [review]
